os/pm/pm_addhistory.c

```diff
@@ -0,0 +1,95 @@
+/****************************************************************************
+ *
+ * Copyright 2020 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <stdlib.h>
+#include <string.h>
+#include <queue.h>
+#include <debug.h>
+#include <time.h>
+#include <assert.h>
+#include <tinyara/pm/pm.h>
+#include "pm.h"
+#include "pm_metrics.h"
+
+#ifdef CONFIG_PM
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: pm_addhistory
+ *
+ * Description:
+ *   This function add history into pm structure.
+ *
+ * Input parameters:
+ *   domain - The PM domain to check
+ *   state - Current PM state
+ *
+ * Returned value:
+ *    None.
+ *
+ ****************************************************************************/
+#ifdef CONFIG_PM_METRICS
+void pm_addhistory(int domain, int state)
+{
+	DEBUGASSERT(domain >= 0 && domain < CONFIG_PM_NDOMAINS && state >= PM_NORMAL && state <= PM_SLEEP);
+	FAR struct pm_domain_s *pdom = &g_pmglobals.domain[domain];
+	struct timespec cur_time;
+	struct pm_statechange_s *node = NULL;
+	struct pm_statechange_s *tail = NULL;
+
+	/* Create an initial state change node with NORMAL state and bootup time */
+
+	node = (struct pm_statechange_s *)pm_alloc(1, sizeof(struct pm_statechange_s));
+	if (node == NULL) {
```

Comment:
minor one you don't have to modify but below is more good habit for future work.
if (a != b) {
    return; //ENOMEM for your case.
}

do something
```

Reply:
Thank you for your comment : )